src/RequestCompressionMiddleware.php

```diff
@@ -0,0 +1,163 @@
+<?php
+namespace Aws;
+
+use GuzzleHttp\Psr7;
+use Psr\Http\Message\RequestInterface;
+
+/**
+ * Used to compress request payloads if the service/operation support it.
+ *
+ * IMPORTANT: this middleware must be added after the "build" step.
+ *
+ * @internal
+ */
+class RequestCompressionMiddleware
+{
+    const REQUEST_MIN_COMPRESSION_SIZE_BYTES = 10240;
+
+    private $api;
+    private $minimumCompressionSize;
+    private $nextHandler;
+    private $encodings;
+    private $encoding;
+    private $encodingMap = [
+        'gzip' => 'gzencode'
+    ];
+
+    /**
+     * Create a middleware wrapper function.
+     *
+     * @return callable
+     */
+    public static function wrap(array $config)
+    {
+        return function (callable $handler) use ($config) {
+            return new self($handler, $config);
+        };
+    }
+
+    public function __construct(callable $nextHandler, $config)
+    {
+        $this->minimumCompressionSize = $this->determineMinimumCompressionSize($config);
+        $this->api = $config['api'];
+        $this->nextHandler = $nextHandler;
+    }
+
+    public function __invoke(CommandInterface $command, RequestInterface $request)
+    {
+        if (isset($command['@request_min_compression_size_bytes'])
+            && is_int($command['@request_min_compression_size_bytes'])
+            && $this->isValidCompressionSize($command['@request_min_compression_size_bytes'])
+        ) {
+            $this->minimumCompressionSize = $command['@request_min_compression_size_bytes'];
+        }
+        $nextHandler = $this->nextHandler;
+        $operation = $this->api->getOperation($command->getName());
+        $requestBodySize = $request->getBody()->getSize();
+        $compressionTrait = isset($operation['requestcompression'])
```

Comment:
shouldn't `requestcompression` be `requestCompression`?

Reply:
It will be modeled as lowercase